### PR TITLE
Define the representation for a tag name.

### PR DESCRIPTION
**Pull request by briandealwis**
A tag is defined in the spec as a manifest identifier, but the spec does not define the acceptable representations for a tag.  This PR provides a definition for the tag name based on the historical description from `docker tag`.

There was no reference to definitions for regular expressions, and other examples used basic regexs, so I didn't use character classes or bounds limits.

Motivated by #52

### CODE REVIEW

File: spec.md [hunk 1]
```diff
@@ -142,6 +142,10 @@ The `<reference>` MUST NOT be in any other format. Throughout this document, `<n
 
 `[a-z0-9]+([._-][a-z0-9]+)*(/[a-z0-9]+([._-][a-z0-9]+)*)*`
 
+Throughout this document, `<reference>` as a tag name MUST be at most 128 characters in length and MUST match the following regular expression:
```
Comment (mikebrow) [inline spec.md:145]: I didn't see the term `tag name` before read right over it.

Please s/tag name/tag/ both here and above on line 140.

**Review by mikebrow — COMMENTED** [review]
see comment regarding `tag name` as a term.

**Comment by briandealwis** [comment]
Oops sorry: these comments ended up in my hidden in my personal email.

@jonjohnsonjr I'm a bit hesitant:
  - **character set:** should be brought in by reference to other standards documents such as HTTP (which defines that the URI and header block must be processed as octets encoded as a superset of US-ASCII).
  - **regex:** there needs to be a reference to a specific form of regexs as POSIX regular expressions don't have `\w`, and RFCs usually use Augmented Backus-Naur Form (ABNF) notation specified in [RFC2234](https://tools.ietf.org/html/rfc2234).  But your regexp showed me that my own regex was incorrect (ok for a tag to start with underscore).
  - **registries should support at least 128 characters:** I'm a bit confused as `128,inf` seems to imply that a registry must support tags that have a minimum of 128 characters?

**Comment by jonjohnsonjr** [comment]
> seems to imply that a registry must support tags that have a minimum of 128 characters?

If registries don't support at least 128 characters in a tag, then they will be "broken" from the perspective of a client that expects to be able to use up to 128 characters.

It's basically Postel's Law: clients need to be conservative in what they do, and registries should be liberal in what they accept.

Is that a good idea? Who knows. The bare minimum, IMO, is that registries should support up to 128 characters. Given that clients are probably not prepared to deal with tags longer than 128 characters, maybe registries should also not support longer tags than that, but it seems a little bit limiting...

**Review by mikebrow — APPROVED** [review]
LGTM

**Comment by jonjohnsonjr** [comment]
LGTM but I'm not a maintainer :)

**Review by waynr — APPROVED** [review]
Looks good to me since it answers a question I brought up in slack. Not sure if I'm a maintainer on this project though -- probably not.

File: spec.md [hunk 2]
```diff
@@ -137,11 +137,15 @@ Typically, the first step in pulling an artifact is to retrieve the manifest. Ho
...
 
-`<name>` refers to the namespace of the repository. `<reference>` MUST be either (a) the digest of the manifest or (b) a tag name.
+`<name>` refers to the namespace of the repository. `<reference>` MUST be either (a) the digest of the manifest or (b) a tag.
 The `<reference>` MUST NOT be in any other format. Throughout this document, `<name>` MUST match the following regular expression:
 
 `[a-z0-9]+([._-][a-z0-9]+)*(/[a-z0-9]+([._-][a-z0-9]+)*)*`
 
+Throughout this document, `<reference>` as a tag MUST be at most 128 characters in length and MUST match the following regular expression:
+
+`[a-zA-Z0-9_]+[a-zA-Z0-9._-]*`
```
Comment (joaodrp) [inline spec.md:147]:
```suggestion
`[a-zA-Z0-9_][a-zA-Z0-9._-]{0,127}`
```

Is there a reason for not adding the quantifier?

For reference, `[\w][\w.-]{0,127}` is what we're using in [distribution/distribution](https://github.com/distribution/distribution). I'm fine with not using `\w`, but I'd keep the length.

**Review by joaodrp — APPROVED** [review]

**Review by sudo-bmitch — APPROVED** [review]

**Comment by vbatts** [comment]
@briandealwis your commit are missing DCO signature

**Comment by briandealwis** [comment]
Added a sign-off.

**Review by vbatts — APPROVED** [review]
🐙

**Comment by briandealwis** [comment]
> Approval required by 1 of: @caniszczyk, @crosbymichael, @dmcgowan, @jdolitsky, @jzelinskie, @mikebrow, @stevvooe

**Review by jzelinskie — APPROVED** [review]
Looks really good in terms of keeping things surgical.

**Pull request merged**